jlm/hls/backend/rvsdg2rhls/rhls-dne.cpp

```diff
@@ -267,6 +269,159 @@ dead_loop(rvsdg::Node * ndmux_node)
+  // one branch
+  if (lcb_node->output(0)->nusers() != 1)
+  {
+    return false;
+  }
+  auto branch_in = dynamic_cast<jlm::rvsdg::node_input *>(*lcb_node->output(0)->begin());
+  auto bo = dynamic_cast<const branch_op *>(&branch_in->node()->GetOperation());
+  if (!branch_in || !bo || !bo->loop)
+  {
+    return false;
+  }
+  // no user
+  if (branch_in->node()->output(1)->nusers())
+  {
+    return false;
+  }
+  // depend on same control
+  auto branch_cond_origin = branch_in->node()->input(0)->origin();
+  auto pred_buf_out = dynamic_cast<jlm::rvsdg::node_output *>(lcb_node->input(0)->origin());
+  if (!pred_buf_out
+      || !dynamic_cast<const predicate_buffer_op *>(&pred_buf_out->node()->GetOperation()))
+  {
+    return false;
+  }
+  auto pred_buf_cond_origin = pred_buf_out->node()->input(0)->origin();
+  // TODO: remove this once predicate buffers decouple combinatorial loops
+  auto extra_buf_out = dynamic_cast<jlm::rvsdg::node_output *>(pred_buf_cond_origin);
+  if (!extra_buf_out || !dynamic_cast<const buffer_op *>(&extra_buf_out->node()->GetOperation()))
+  {
+    return false;
+  }
+  auto extra_buf_cond_origin = extra_buf_out->node()->input(0)->origin();
+
+  if (auto pred_be = dynamic_cast<backedge_argument *>(extra_buf_cond_origin))
+  {
+    extra_buf_cond_origin = pred_be->result()->origin();
+  }
+  if (extra_buf_cond_origin != branch_cond_origin)
+  {
+    return false;
+  }
+  // divert users
+  branch_in->node()->output(0)->divert_users(lcb_node->input(1)->origin());
+  remove(branch_in->node());
+  remove(lcb_node);
+  return true;
+}
+
+bool
+fix_mem_split(rvsdg::Node * split_node)
+{
+  if (split_node->noutputs() == 1)
+  {
+    split_node->output(0)->divert_users(split_node->input(0)->origin());
+    JLM_ASSERT(split_node->IsDead());
+    remove(split_node);
+    return true;
+  }
+  // this merges downward and removes unused outputs (should only exist as a result of eliminating
+  // merges)
+  std::vector<rvsdg::output *> combined_outputs;
+  for (size_t i = 0; i < split_node->noutputs(); ++i)
+  {
+    if (split_node->output(i)->IsDead())
+      continue;
+    auto user = get_mem_state_user(split_node->output(i));
+    if (TryGetOwnerOp<llvm::MemoryStateSplitOperation>(*user))
+    {
+      auto sub_split = rvsdg::TryGetOwnerNode<rvsdg::SimpleNode>(*user);
+      for (size_t j = 0; j < sub_split->noutputs(); ++j)
+      {
+        combined_outputs.push_back(sub_split->output(j));
+      }
+    }
+    else
+    {
+      combined_outputs.push_back(split_node->output(i));
+    }
+  }
+  if (combined_outputs.size() != split_node->noutputs())
+  {
+    auto new_outputs = llvm::MemoryStateSplitOperation::Create(
+        *split_node->input(0)->origin(),
+        combined_outputs.size());
+    for (size_t i = 0; i < combined_outputs.size(); ++i)
+    {
+      combined_outputs[i]->divert_users(new_outputs[i]);
+    }
+    return true;
+  }
+  return false;
+}
+
+bool
+fix_mem_merge(rvsdg::Node * merge_node)
```

Comment:
Removes merges that only have one state edge.
Replaces multiple parallel state edges with a single edge.
Combines multiple merge nodes into a single node.

Reply:
Should likely be part of a node reduction pass instead of DNE